unix/w0vncserver/GSocketSource.cxx

```diff
@@ -0,0 +1,248 @@
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <assert.h>
+
+#include <glib.h>
+
+#include <network/TcpSocket.h>
+#include <core/LogWriter.h>
+#include <rfb/VNCServerST.h>
+#include <rdr/FdOutStream.h>
+
+#include <map>
+
+#include "GSocketSource.h"
+
+static core::LogWriter vlog("GSocketMonitor");
+
+struct SocketState {
+  void* tag;
+  bool prevHadBufferedData;
+};
+
+struct _GSocketSource {
+  GSource base;
+  GSocketSource* instance;
+};
+
+struct ListenerReadyEvent {
+  GSocketSource* instance;
+  network::SocketListener* listener;
+  GIOCondition condition;
+};
+
+GSourceFuncs GSocketSource::sourceFuncs {
+  .prepare = [](GSource* source, int* timeout) {
+    GSocketSource* self = ((_GSocketSource*)(source))->instance;
+    return self->prepare(timeout);
+  },
+  .check = nullptr,
+  .dispatch = [](GSource* source, GSourceFunc callback, void* userData) {
+    GSocketSource* self = ((_GSocketSource*)(source))->instance;
+    return self->dispatch(callback, userData);
+  },
+  .finalize = nullptr,
+  .closure_callback = nullptr,
+  .closure_marshal = nullptr
+};
+
+GSocketSource::GSocketSource(rfb::VNCServer* server,
+                             std::list<network::SocketListener*> *listeners)
+  : source(nullptr), server_(server), listeners_(listeners)
+{
+  source = (_GSocketSource*)g_source_new(&sourceFuncs, sizeof(_GSocketSource));
+  previousCondition = G_IO_IN;
+
+  source->instance = this;
+}
+
+GSocketSource::~GSocketSource()
+{
+  std::list<network::Socket*> sockets;
+
+  for (GIOChannel* channel : channels)
+    g_io_channel_unref(channel);
+
+  for (ListenerReadyEvent* event : listenerEvents)
+    delete event;
+
+  server_->getSockets(&sockets);
+  for (network::Socket* sock : sockets) {
+    server_->removeSocket(sock);
+    delete sock;
+  }
+
+  // GLib will take care of the monitored FDs on our source
+  g_source_unref(&source->base);
+  g_source_destroy(&source->base);
+}
+
+void GSocketSource::attach(GMainContext* context)
+{
+  assert(source);
+
+  g_source_attach((GSource*)source, context);
+}
+
+void GSocketSource::listen()
+{
+  assert(listeners_->size());
+
+  for (network::SocketListener* listener : *listeners_) {
+    ListenerReadyEvent* event;
+    GIOChannel* channel;
+    int fd;
+
+    fd = listener->getFd();
+
+    channel = g_io_channel_unix_new(fd);
+    channels.push_back(channel);
+
+    event = new ListenerReadyEvent();
+    event->instance = this;
+    event->listener = listener;
+
+    listenerEvents.push_back(event);
+
+    g_io_add_watch(channel,
+                   static_cast<GIOCondition>((G_IO_IN | G_IO_ERR | G_IO_HUP)),
+                   [](GIOChannel*, GIOCondition condition, gpointer data) {
+      ListenerReadyEvent* event_ = static_cast<ListenerReadyEvent*>(data);
```

Comment:
The `_` doesn't seem necessary here?

Reply:
Well that's annoying. But if I search the issue, there is apparently some argument for warning about that.

I guess we'll have to live with it.